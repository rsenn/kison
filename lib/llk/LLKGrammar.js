const Grammar = require('../Grammar');
const parseCode = require('./parse');
const { serializeObject } = require('../utils');

class LLKGrammar extends Grammar {
  buildMeta() {}

  buildProductions() {
    const firstProduction = this.productions[0];
    this.productions.splice(0, 1);
    this.eliminateLeftRecursive();
    this.productions.splice(0, 0, firstProduction);
  }

  expandProductionsInternal() {
    this.expandOneOrMoreSymbol();
    this.expandProductionPriority();
  }

  genCodeInternal(code) {
    for (const key of Object.keys(parseCode)) {
      code.push(`var ${key} = ${serializeObject(parseCode[key])};`);
    }
    code.push(`${parseCode.initLLK.name}();`);
    code.push(`parser.parse = ${parseCode.parse.name};`);
    return code.join('\n');
  }
}

module.exports = LLKGrammar;
