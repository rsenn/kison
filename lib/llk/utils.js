// @ts-check

const data = require('../data');

let { cachedStateMatchMap } = data;

function setStateCacheMatch(state, tokensLength) {
  return
  let holder = cachedStateMatchMap.get(state);
  if (!holder) {
    holder = new Set();
    cachedStateMatchMap.set(state, holder);
  }
  holder.add(tokensLength);
}

function getStateCacheMatch(state, tokensLength) {
  return;
  // consumed token length may be different!
  const holder = cachedStateMatchMap.get(state);
  return holder?.has(tokensLength);
}

function clearStateMatchCache() {
  return
  cachedStateMatchMap = new Map();
}

module.exports = {
  setStateCacheMatch,
  getStateCacheMatch,
  clearStateMatchCache,
};