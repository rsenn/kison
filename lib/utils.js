/**
 * @ignore
 * utils for kison.
 * @author yiminghe@gmail.com
 */

var util = require("modulex-util");

var doubleReg = /"/g;
var single = /'/g;

const gened = {
  mix(to, from) {
    for (var f in from) {
      to[f] = from[f];
    }
  },

  isArray(obj) {
    return "[object Array]" === Object.prototype.toString.call(obj);
  },

  inArray(item, arr) {
    for (var i = 0, l = arr.length; i < l; i++) {
      if (arr[i] === item) {
        return true;
      }
    }
    return false;
  },

  each(object, fn, context) {
    if (object) {
      var key,
        val,
        length,
        i = 0;

      context = context || null;

      if (!isArray(object)) {
        for (key in object) {
          // can not use hasOwnProperty
          if (fn.call(context, object[key], key, object) === false) {
            break;
          }
        }
      } else {
        length = object.length;
        for (val = object[0]; i < length; val = object[++i]) {
          if (fn.call(context, val, i, object) === false) {
            break;
          }
        }
      }
    }
  }
};

var { isArray } = gened;

const utils = (module.exports = Object.assign({}, util, gened, {
  eachRhs(rhs, fn) {
    utils.each(rhs, (r, index) => {
      if (typeof r !== "string") {
        return;
      }
      fn(r, index);
    });
  },

  assertSymbolString(symbol) {
    if (typeof symbol !== "string") {
      throw new Error("invalid symbol".symbol);
    }
  },

  filterRhs(rhs) {
    return rhs.filter(r => typeof r === "string");
  },

  toFunctionString(fn, name) {
    const str = fn.toString();
    if (str.match(/^function\s+\(/)) {
      return str.replace(/^function\s+\(/, `function ${name || fn.name}(`);
    }
    return "function " + str;
  },

  escapeString(str, quote) {
    var regexp = single;
    if (quote === '"') {
      regexp = doubleReg;
    } else {
      quote = "'";
    }
    return str
      .replace(/\\/g, "\\\\")
      .replace(/\r/g, "\\r")
      .replace(/\n/g, "\\n")
      .replace(/\t/g, "\\t")
      .replace(regexp, "\\" + quote);
  },

  serializeObject(obj, excludeReg) {
    var r;

    if (
      excludeReg &&
      typeof excludeReg === "function" &&
      (r = excludeReg(obj)) === false
    ) {
      return false;
    }

    if (r !== undefined) {
      obj = r;
    }

    var ret = [];

    if (typeof obj === "string") {
      return "'" + utils.escapeString(obj) + "'";
    } else if (typeof obj === "number") {
      return obj + "";
    } else if (util.isRegExp(obj)) {
      return (
        "/" +
        obj.source +
        "/" +
        (obj.global ? "g" : "") +
        (obj.ignoreCase ? "i" : "") +
        (obj.multiline ? "m" : "")
      );
    } else if (util.isArray(obj)) {
      ret.push("[");
      var sub = [];
      util.each(obj, function(v) {
        var t = utils.serializeObject(v, excludeReg);
        if (t !== false) {
          sub.push(t);
        }
      });
      ret.push(sub.join(", "));
      ret.push("]");
      return ret.join("");
    } else if (typeof obj === "object") {
      ret = [];
      ret[0] = "{";
      var start = 1;
      for (var i in obj) {
        var v = obj[i];
        if (excludeReg && util.isRegExp(excludeReg) && i.match(excludeReg)) {
          continue;
        }
        var t = utils.serializeObject(v, excludeReg);
        if (t === false) {
          continue;
        }
        /*jshint quotmark:false*/
        var key = "'" + utils.escapeString(i) + "'";
        ret.push((start ? "" : ",") + key + ": " + t);
        start = 0;
      }
      ret.push("}");
      return ret.join("\n");
    } else if (typeof obj === "function") {
      return (obj + "").replace(/^[\w-]+/, "function");
    } else {
      return obj + "";
    }
  }
}));
