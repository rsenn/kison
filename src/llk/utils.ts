// @ts-check

import data from '../data';

let { productionsBySymbol } = data;

function isSymbol(s: string) {
  return !!productionsBySymbol[s];
}

export { isSymbol };
