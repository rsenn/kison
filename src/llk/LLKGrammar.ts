import Grammar from '../Grammar';
import * as parseCode from './parse';
import utils from '../utils';

const { serializeObject } = utils;

class LLKGrammar extends Grammar {
  buildMeta() {}

  buildProductions() {
    const firstProduction = this.productionInstances[0];
    this.productionInstances.splice(0, 1);
    this.eliminateLeftRecursive();
    this.productionInstances.splice(0, 0, firstProduction);
  }

  expandProductionsInternal() {
    this.expandOneOrMoreSymbol();
    this.expandProductionPriority();
  }

  genCodeInternal(code: string[]) {
    for (const key of Object.keys(parseCode)) {
      code.push(`var ${key} = ${serializeObject((parseCode as any)[key])};`);
    }
    code.push(`${parseCode.initLLK.name}();`);
    code.push(`parser.parse = ${parseCode.parse.name};`);
    return code.join('\n');
  }
}

export default LLKGrammar;
