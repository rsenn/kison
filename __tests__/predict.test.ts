
import LLKGrammar from '../src/llk/LLKGrammar';
import { prettyJson, run, HIDDEN_LEXER_RULE } from './utils';
import Lexer from '../src/Lexer';
import type Parser from '../src/parser';

const startGroupMarker = `'('`;
const startPredictGroupMarker = `'(?'`;
const endGroupMarker = `')'`;
const alterMark = `'|'`;
const alterPredictMark = `'|?'`;

describe('llk predict', () => {
  it('simple predict', () => {
    interface ThisArg {
      userData: {
        count?: number;
      };
      lexer: Lexer;
    }
    var grammar = new LLKGrammar({
      productions: [
        {
          symbol: 'program',
          rhs: ['sequence+'],
        },

        {
          symbol: 'sequence',
          rhs: ['NUMBER', 'sequence2'],
          action(this: ThisArg) {
            this.userData = {};
          },
        },
        {
          symbol: 'sequence2',
          rhs: [
            'NUMBER',
            function (this: ThisArg) {
              if (this.userData.count) this.userData.count--;
            },
            'sequence2',
          ],
          predict(this: ThisArg) {
            if (this.userData.count === undefined) {
              const tokens = this.lexer.tokens;
              let index = tokens.length - 2;
              let token = tokens[index];
              while (index && token.channel) {
                token = tokens[--index];
              }
              this.userData.count = parseInt(token.text);
            } else if (this.userData.count === 0) {
              return false;
            }
          },
        },
        {
          symbol: 'sequence2',
          rhs: [],
          predict(this: any) {
            if (this.userData.count) {
              return false;
            }
          },
        },
      ],
      lexer: {
        rules: [
          HIDDEN_LEXER_RULE,
          {
            token: 'NUMBER',
            regexp: /\d+/,
          },
        ],
      },
    });

    const code = grammar.genCode();
    const parser = run(code);
    let ret = parser.parse('2 1 1 3 1 1 1');
    expect(ret.error).toBeFalsy();
    expect(prettyJson(ret.ast)).toMatchSnapshot();
    ret = parser.parse('2 1 1 3 1 1 ');
    expect(ret.error).toMatchSnapshot();
  });


  it('support group predict', () => {
    var grammar = new LLKGrammar({
      productions: [
        {
          symbol: 'program',
          rhs: ['sequence+'],
        },

        {
          symbol: 'sequence',
          rhs: ['NUMBER',

            startPredictGroupMarker,

            function (this:typeof Parser){
              const token = this.lexer.getLastToken((t)=>!t.channel);
              return token.text==='3';
            },
            'sequence3',
            alterPredictMark,
            function (this:typeof Parser){
              const token = this.lexer.getLastToken((t)=>!t.channel);
              return token.text==='2';
            },
            'sequence2',
            alterMark,
            'sequence1',
            endGroupMarker,
          ]
        },
        {
          symbol: 'sequence1',
          rhs: [
            'NUMBER',
          ],
        },
        {
          symbol: 'sequence2',
          rhs: ['NUMBER', 'NUMBER',],
        },
        {
          symbol: 'sequence3',
          rhs: ['NUMBER', 'NUMBER', 'NUMBER'],
        },
      ],
      lexer: {
        rules: [
          HIDDEN_LEXER_RULE,
          {
            token: 'NUMBER',
            regexp: /\d+/,
          },
        ],
      },
    });


    const code = grammar.genCode();
    const parser = run(code);
    let ret = parser.parse('2 1 1 3 1 1 1');
    expect(ret.error).toBeFalsy();
    expect(prettyJson(ret.ast)).toMatchSnapshot();
    ret = parser.parse('2 1 1 3 1 1 ');
    expect(ret.error).toMatchSnapshot();
    // global match!
    ret = parser.parse('2 1 1 5');
    expect(ret.error).toMatchSnapshot();
  });
});