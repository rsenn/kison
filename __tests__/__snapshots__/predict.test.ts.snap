// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`llk predict simple predict 1`] = `
"{
  'start': 0,
  'end': 0,
  'firstLine': 1,
  'lastLine': 0,
  'firstColumn': 1,
  'lastColumn': 0,
  'symbol': 'program',
  'type': 'symbol',
  'children': [
    {
      'start': 0,
      'end': 0,
      'firstLine': 1,
      'lastLine': 0,
      'firstColumn': 1,
      'lastColumn': 0,
      'symbol': 'sequence',
      'type': 'symbol',
      'children': [
        {
          'start': 0,
          'end': 1,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 1,
          'lastColumn': 2,
          'token': 'NUMBER',
          'type': 'token',
          'text': '2'
        },
        {
          'start': 2,
          'end': 0,
          'firstLine': 1,
          'lastLine': 0,
          'firstColumn': 3,
          'lastColumn': 0,
          'symbol': 'sequence2',
          'type': 'symbol',
          'children': [
            {
              'start': 2,
              'end': 3,
              'firstLine': 1,
              'lastLine': 1,
              'firstColumn': 3,
              'lastColumn': 4,
              'token': 'NUMBER',
              'type': 'token',
              'text': '1'
            },
            {
              'start': 4,
              'end': 0,
              'firstLine': 1,
              'lastLine': 0,
              'firstColumn': 5,
              'lastColumn': 0,
              'symbol': 'sequence2',
              'type': 'symbol',
              'children': [
                {
                  'start': 4,
                  'end': 5,
                  'firstLine': 1,
                  'lastLine': 1,
                  'firstColumn': 5,
                  'lastColumn': 6,
                  'token': 'NUMBER',
                  'type': 'token',
                  'text': '1'
                },
                {
                  'start': 0,
                  'end': 0,
                  'firstLine': 0,
                  'lastLine': 0,
                  'firstColumn': 0,
                  'lastColumn': 0,
                  'symbol': 'sequence2',
                  'type': 'symbol',
                  'children': [],
                  'ruleIndex': 4,
                  'internalRuleIndex': 4,
                  'id': 6
                }
              ],
              'ruleIndex': 3,
              'internalRuleIndex': 3,
              'id': 5
            }
          ],
          'ruleIndex': 3,
          'internalRuleIndex': 3,
          'id': 4
        }
      ],
      'ruleIndex': 2,
      'internalRuleIndex': 2,
      'id': 3
    },
    {
      'start': 6,
      'end': 0,
      'firstLine': 1,
      'lastLine': 0,
      'firstColumn': 7,
      'lastColumn': 0,
      'symbol': 'sequence',
      'type': 'symbol',
      'children': [
        {
          'start': 6,
          'end': 7,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 7,
          'lastColumn': 8,
          'token': 'NUMBER',
          'type': 'token',
          'text': '3'
        },
        {
          'start': 8,
          'end': 0,
          'firstLine': 1,
          'lastLine': 0,
          'firstColumn': 9,
          'lastColumn': 0,
          'symbol': 'sequence2',
          'type': 'symbol',
          'children': [
            {
              'start': 8,
              'end': 9,
              'firstLine': 1,
              'lastLine': 1,
              'firstColumn': 9,
              'lastColumn': 10,
              'token': 'NUMBER',
              'type': 'token',
              'text': '1'
            },
            {
              'start': 10,
              'end': 0,
              'firstLine': 1,
              'lastLine': 0,
              'firstColumn': 11,
              'lastColumn': 0,
              'symbol': 'sequence2',
              'type': 'symbol',
              'children': [
                {
                  'start': 10,
                  'end': 11,
                  'firstLine': 1,
                  'lastLine': 1,
                  'firstColumn': 11,
                  'lastColumn': 12,
                  'token': 'NUMBER',
                  'type': 'token',
                  'text': '1'
                },
                {
                  'start': 12,
                  'end': 0,
                  'firstLine': 1,
                  'lastLine': 0,
                  'firstColumn': 13,
                  'lastColumn': 0,
                  'symbol': 'sequence2',
                  'type': 'symbol',
                  'children': [
                    {
                      'start': 12,
                      'end': 13,
                      'firstLine': 1,
                      'lastLine': 1,
                      'firstColumn': 13,
                      'lastColumn': 14,
                      'token': 'NUMBER',
                      'type': 'token',
                      'text': '1'
                    },
                    {
                      'start': 0,
                      'end': 0,
                      'firstLine': 0,
                      'lastLine': 0,
                      'firstColumn': 0,
                      'lastColumn': 0,
                      'symbol': 'sequence2',
                      'type': 'symbol',
                      'children': [],
                      'ruleIndex': 4,
                      'internalRuleIndex': 4,
                      'id': 11
                    }
                  ],
                  'ruleIndex': 3,
                  'internalRuleIndex': 3,
                  'id': 10
                }
              ],
              'ruleIndex': 3,
              'internalRuleIndex': 3,
              'id': 9
            }
          ],
          'ruleIndex': 3,
          'internalRuleIndex': 3,
          'id': 8
        }
      ],
      'ruleIndex': 2,
      'internalRuleIndex': 2,
      'id': 7
    }
  ],
  'ruleIndex': 1,
  'internalRuleIndex': 1,
  'id': 2
}"
`;

exports[`llk predict simple predict 2`] = `
Object {
  "errorMessage": "syntax error at line 1:
2 1 1 3 1 1 
------------^
'NUMBER' expected.
current token: '$EOF'.",
  "expected": Array [
    "NUMBER",
  ],
  "recovery": false,
  "symbol": "sequence2",
  "tip": "'NUMBER' expected.
current token: '$EOF'.",
  "token": Object {
    "end": 12,
    "firstColumn": 13,
    "firstLine": 1,
    "lastColumn": 13,
    "lastLine": 1,
    "start": 12,
    "t": "$EOF",
    "text": "",
    "token": "$EOF",
  },
}
`;

exports[`llk predict support group predict 1`] = `
"{
  'start': 0,
  'end': 13,
  'firstLine': 1,
  'lastLine': 1,
  'firstColumn': 1,
  'lastColumn': 14,
  'symbol': 'program',
  'type': 'symbol',
  'children': [
    {
      'start': 0,
      'end': 5,
      'firstLine': 1,
      'lastLine': 1,
      'firstColumn': 1,
      'lastColumn': 6,
      'symbol': 'sequence',
      'type': 'symbol',
      'children': [
        {
          'start': 0,
          'end': 1,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 1,
          'lastColumn': 2,
          'token': 'NUMBER',
          'type': 'token',
          'text': '2'
        },
        {
          'start': 2,
          'end': 5,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 3,
          'lastColumn': 6,
          'symbol': 'sequence2',
          'type': 'symbol',
          'children': [
            {
              'start': 2,
              'end': 3,
              'firstLine': 1,
              'lastLine': 1,
              'firstColumn': 3,
              'lastColumn': 4,
              'token': 'NUMBER',
              'type': 'token',
              'text': '1'
            },
            {
              'start': 4,
              'end': 5,
              'firstLine': 1,
              'lastLine': 1,
              'firstColumn': 5,
              'lastColumn': 6,
              'token': 'NUMBER',
              'type': 'token',
              'text': '1'
            }
          ],
          'ruleIndex': 4,
          'internalRuleIndex': 10,
          'id': 4
        }
      ],
      'ruleIndex': 2,
      'internalRuleIndex': 8,
      'id': 3
    },
    {
      'start': 6,
      'end': 13,
      'firstLine': 1,
      'lastLine': 1,
      'firstColumn': 7,
      'lastColumn': 14,
      'symbol': 'sequence',
      'type': 'symbol',
      'children': [
        {
          'start': 6,
          'end': 7,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 7,
          'lastColumn': 8,
          'token': 'NUMBER',
          'type': 'token',
          'text': '3'
        },
        {
          'start': 8,
          'end': 13,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 9,
          'lastColumn': 14,
          'symbol': 'sequence3',
          'type': 'symbol',
          'children': [
            {
              'start': 8,
              'end': 9,
              'firstLine': 1,
              'lastLine': 1,
              'firstColumn': 9,
              'lastColumn': 10,
              'token': 'NUMBER',
              'type': 'token',
              'text': '1'
            },
            {
              'start': 10,
              'end': 11,
              'firstLine': 1,
              'lastLine': 1,
              'firstColumn': 11,
              'lastColumn': 12,
              'token': 'NUMBER',
              'type': 'token',
              'text': '1'
            },
            {
              'start': 12,
              'end': 13,
              'firstLine': 1,
              'lastLine': 1,
              'firstColumn': 13,
              'lastColumn': 14,
              'token': 'NUMBER',
              'type': 'token',
              'text': '1'
            }
          ],
          'ruleIndex': 5,
          'internalRuleIndex': 11,
          'id': 6
        }
      ],
      'ruleIndex': 2,
      'internalRuleIndex': 8,
      'id': 5
    }
  ],
  'ruleIndex': 1,
  'internalRuleIndex': 1,
  'id': 2
}"
`;

exports[`llk predict support group predict 2`] = `
"{
  'start': 0,
  'end': 11,
  'firstLine': 1,
  'lastLine': 1,
  'firstColumn': 1,
  'lastColumn': 12,
  'symbol': 'program',
  'type': 'symbol',
  'children': [
    {
      'start': 0,
      'end': 3,
      'firstLine': 1,
      'lastLine': 1,
      'firstColumn': 1,
      'lastColumn': 4,
      'symbol': 'sequence',
      'type': 'symbol',
      'children': [
        {
          'start': 0,
          'end': 1,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 1,
          'lastColumn': 2,
          'token': 'NUMBER',
          'type': 'token',
          'text': '2'
        },
        {
          'start': 2,
          'end': 3,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 3,
          'lastColumn': 4,
          'symbol': 'sequence1',
          'type': 'symbol',
          'children': [
            {
              'start': 2,
              'end': 3,
              'firstLine': 1,
              'lastLine': 1,
              'firstColumn': 3,
              'lastColumn': 4,
              'token': 'NUMBER',
              'type': 'token',
              'text': '1'
            }
          ],
          'ruleIndex': 3,
          'internalRuleIndex': 9,
          'id': 4
        }
      ],
      'ruleIndex': 2,
      'internalRuleIndex': 8,
      'id': 3
    },
    {
      'start': 4,
      'end': 7,
      'firstLine': 1,
      'lastLine': 1,
      'firstColumn': 5,
      'lastColumn': 8,
      'symbol': 'sequence',
      'type': 'symbol',
      'children': [
        {
          'start': 4,
          'end': 5,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 5,
          'lastColumn': 6,
          'token': 'NUMBER',
          'type': 'token',
          'text': '1'
        },
        {
          'start': 6,
          'end': 7,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 7,
          'lastColumn': 8,
          'symbol': 'sequence1',
          'type': 'symbol',
          'children': [
            {
              'start': 6,
              'end': 7,
              'firstLine': 1,
              'lastLine': 1,
              'firstColumn': 7,
              'lastColumn': 8,
              'token': 'NUMBER',
              'type': 'token',
              'text': '3'
            }
          ],
          'ruleIndex': 3,
          'internalRuleIndex': 9,
          'id': 6
        }
      ],
      'ruleIndex': 2,
      'internalRuleIndex': 8,
      'id': 5
    },
    {
      'start': 8,
      'end': 11,
      'firstLine': 1,
      'lastLine': 1,
      'firstColumn': 9,
      'lastColumn': 12,
      'symbol': 'sequence',
      'type': 'symbol',
      'children': [
        {
          'start': 8,
          'end': 9,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 9,
          'lastColumn': 10,
          'token': 'NUMBER',
          'type': 'token',
          'text': '1'
        },
        {
          'start': 10,
          'end': 11,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 11,
          'lastColumn': 12,
          'symbol': 'sequence1',
          'type': 'symbol',
          'children': [
            {
              'start': 10,
              'end': 11,
              'firstLine': 1,
              'lastLine': 1,
              'firstColumn': 11,
              'lastColumn': 12,
              'token': 'NUMBER',
              'type': 'token',
              'text': '1'
            }
          ],
          'ruleIndex': 3,
          'internalRuleIndex': 9,
          'id': 8
        }
      ],
      'ruleIndex': 2,
      'internalRuleIndex': 8,
      'id': 7
    }
  ],
  'ruleIndex': 1,
  'internalRuleIndex': 1,
  'id': 2
}"
`;

exports[`llk predict support group predict 3`] = `
"{
  'start': 0,
  'end': 7,
  'firstLine': 1,
  'lastLine': 1,
  'firstColumn': 1,
  'lastColumn': 8,
  'symbol': 'program',
  'type': 'symbol',
  'children': [
    {
      'start': 0,
      'end': 3,
      'firstLine': 1,
      'lastLine': 1,
      'firstColumn': 1,
      'lastColumn': 4,
      'symbol': 'sequence',
      'type': 'symbol',
      'children': [
        {
          'start': 0,
          'end': 1,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 1,
          'lastColumn': 2,
          'token': 'NUMBER',
          'type': 'token',
          'text': '2'
        },
        {
          'start': 2,
          'end': 3,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 3,
          'lastColumn': 4,
          'symbol': 'sequence1',
          'type': 'symbol',
          'children': [
            {
              'start': 2,
              'end': 3,
              'firstLine': 1,
              'lastLine': 1,
              'firstColumn': 3,
              'lastColumn': 4,
              'token': 'NUMBER',
              'type': 'token',
              'text': '1'
            }
          ],
          'ruleIndex': 3,
          'internalRuleIndex': 9,
          'id': 4
        }
      ],
      'ruleIndex': 2,
      'internalRuleIndex': 8,
      'id': 3
    },
    {
      'start': 4,
      'end': 7,
      'firstLine': 1,
      'lastLine': 1,
      'firstColumn': 5,
      'lastColumn': 8,
      'symbol': 'sequence',
      'type': 'symbol',
      'children': [
        {
          'start': 4,
          'end': 5,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 5,
          'lastColumn': 6,
          'token': 'NUMBER',
          'type': 'token',
          'text': '1'
        },
        {
          'start': 6,
          'end': 7,
          'firstLine': 1,
          'lastLine': 1,
          'firstColumn': 7,
          'lastColumn': 8,
          'symbol': 'sequence1',
          'type': 'symbol',
          'children': [
            {
              'start': 6,
              'end': 7,
              'firstLine': 1,
              'lastLine': 1,
              'firstColumn': 7,
              'lastColumn': 8,
              'token': 'NUMBER',
              'type': 'token',
              'text': '5'
            }
          ],
          'ruleIndex': 3,
          'internalRuleIndex': 9,
          'id': 6
        }
      ],
      'ruleIndex': 2,
      'internalRuleIndex': 8,
      'id': 5
    }
  ],
  'ruleIndex': 1,
  'internalRuleIndex': 1,
  'id': 2
}"
`;
